### PR TITLE
fixed problems when working with msspoly and the rigidbodymanipulator

### PR DESCRIPTION
**Pull request by tarquasso**

### CODE REVIEW

File: systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m [hunk 1]
```diff
@@ -1062,8 +1062,10 @@
       end
       
       model = applyToAllRigidBodyElements(model,'updateParams',fr.getPoly,p);
-
-      model = compile(model);
+      
+      if isnumeric(p)
+        model = compile(model);
+      end
```
Comment (RussTedrake) [inline systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m]: i worry that we might be missing important steps here.  what failed exactly in the compile for non-numeric?  (maybe we can leave something out more selectively)?

File: systems/plants/RigidBody.m [hunk 2]
```diff
@@ -310,7 +310,11 @@ function dofnum(obj)
       % class. (maybe I should move it up to there?)
       fn = fieldnames(body.param_bindings);
       for i=1:length(fn)
-        body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
+        if isnumeric(pval)
+          body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
```
Comment (RussTedrake) [inline systems/plants/RigidBody.m]: the double was only there to make things numeric if they were not.  clearly the original code was not written to support using setParams to pass in a substitution of symbolic variables.  is that what you're trying to accomplish?

File: systems/plants/RigidBodyTorsionalSpring.m [hunk 3]
```diff
@@ -14,7 +14,7 @@
     function [f_ext,df_ext] = computeSpatialForce(obj,manip,q,qd)
       theta = q(manip.body(obj.child_body).position_num);
       torque = obj.k*(obj.rest_angle - theta);
-      f_ext = sparse(6,getNumBodies(manip));
+      f_ext = q(1)*sparse(6,getNumBodies(manip));
```
Comment (RussTedrake) [inline systems/plants/RigidBodyTorsionalSpring.m]: this looks totally fine.

File: systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m (hunk 1 above)
Comment (tarquasso) [inline systems/plants/@RigidBodyManipulator/RigidBodyManipulator.m]: When running this:
r1 = RigidBodyManipulator('tensionWParams.urdf'); %tensionWParams.urdf is a parameterized urdf model

q=msspoly('q',nq);
s=msspoly('s',nq);
c=msspoly('c',nq);
qt=TrigPoly(q,s,c);
qd=msspoly('qd',nq);
qdd=msspoly('qdd',nq);
u=msspoly('u',nu);
p=r1.getParamFrame.getPoly;
pr1 = setParams(r1,p);

get this error when runnin 'pr1 = setParams(r1,p);':

Error using constructModelmex
mxGetPr can only be called on arguments which correspond to Matlab doubles

Error in RigidBodyManipulator/createMexPointer (line 2084)
        obj.mex_model_ptr = constructModelmex(obj);

Error in RigidBodyManipulator/compile (line 818)
      model = createMexPointer(model);

Error in RigidBodyManipulator/setParams (line 1067)
        model = compile(model);

Error in extractDynParams (line 16)
pr1 = setParams(r1,p);

File: systems/plants/RigidBody.m (hunk 2 above)
Comment (tarquasso) [inline systems/plants/RigidBody.m]: yes, trying to substitute symbolic variables.

Here is the code snippet:
r1 = RigidBodyManipulator('tensionWParams.urdf');
q=msspoly('q',nq);
s=msspoly('s',nq);
c=msspoly('c',nq);
qt=TrigPoly(q,s,c);
qd=msspoly('qd',nq);
qdd=msspoly('qdd',nq);
u=msspoly('u',nu);
p=r1.getParamFrame.getPoly;
pr1 = setParams(r1,p);

Throws this error:
Error using msspoly/double (line 21)
Cannot cast non-constant msspoly to double.

Error in RigidBody/updateParams (line 314)
          body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));

Error in RigidBodyManipulator/applyToAllRigidBodyElements (line 2441)
        model.body(i) = feval(fcn,model.body(i),varargin{:});

Error in RigidBodyManipulator/setParams (line 1064)
      model = applyToAllRigidBodyElements(model,'updateParams',fr.getPoly,p);

Error in extractDynParams (line 16)
pr1 = setParams(r1,p);

Comment (tarquasso) [inline systems/plants/RigidBody.m]: same problem in line 35 of plants/RigidBodyElement.m:

       fn = fieldnames(body.param_bindings);
       for i=1:length(fn)
-        body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
+        if isnumeric(pval)
+          body.(fn{i}) = double(subs(body.param_bindings.(fn{i}),poly,pval));
+        else
+          body.(fn{i}) = subs(body.param_bindings.(fn{i}),poly,pval);
+        end

File: systems/plants/RigidBodyTorsionalSpring.m (hunk 3 above)
Comment (tarquasso) [inline systems/plants/RigidBodyTorsionalSpring.m]: same problem in line 348 of RigidBodyManipulator.m:

damping = zeros(nv, 1);

should be changed to:
       
damping = v(1)*zeros(nv, 1);

**Comment by tarquasso** [comment]
added line notes to explain the 3 code changes.


On May 16, 2015, at 09:43, Russ Tedrake <notifications@github.com> wrote:

Assigned #1104 <https://github.com/RobotLocomotion/drake/pull/1104> to @tarquasso <https://github.com/tarquasso>.

**Comment by RussTedrake** [comment]
i think i understand everything you commented on, and will take a look to try to pull it in soon.

**Comment by RussTedrake** [comment]
i've merged this into master and fixed the overly aggressive "don't compile" part.  will send a new pull request with the merge.

**Pull request closed without merging**